src/mem/impls.rs

```diff
@@ -1,27 +1,221 @@
+use core::intrinsics::likely;
+
+const WORD_SIZE: usize = core::mem::size_of::<usize>();
+const WORD_MASK: usize = WORD_SIZE - 1;
+
+const WORD_COPY_THRESHOLD: usize = if 2 * WORD_SIZE > 16 {
+    2 * WORD_SIZE
+} else {
+    16
+};
+
+#[inline(always)]
+unsafe fn copy_forward_bytes(mut dest: *mut u8, mut src: *const u8, n: usize) {
+    let dest_end = dest.add(n);
+    while dest < dest_end {
+        *dest = *src;
+        dest = dest.add(1);
+        src = src.add(1);
+    }
+}
+
 #[inline(always)]
-pub unsafe fn copy_forward(dest: *mut u8, src: *const u8, n: usize) {
-    let mut i = 0;
-    while i < n {
-        *dest.add(i) = *src.add(i);
-        i += 1;
+unsafe fn copy_forward_aligned_words(dest: *mut u8, src: *const u8, n: usize) {
+    let mut dest_usize = dest as *mut usize;
+    let mut src_usize = src as *mut usize;
+    let dest_end = dest.add(n) as *mut usize;
+
+    while dest_usize < dest_end {
+        *dest_usize = *src_usize;
+        dest_usize = dest_usize.add(1);
+        src_usize = src_usize.add(1);
     }
 }
 
 #[inline(always)]
-pub unsafe fn copy_backward(dest: *mut u8, src: *const u8, n: usize) {
-    // copy from end
-    let mut i = n;
-    while i != 0 {
-        i -= 1;
-        *dest.add(i) = *src.add(i);
+unsafe fn copy_forward_misaligned_words(dest: *mut u8, src: *const u8, n: usize) {
+    let mut dest_usize = dest as *mut usize;
+    let dest_end = dest.add(n) as *mut usize;
+
+    // Calculate the misalignment offset and shift needed to reassemble value.
+    let offset = src as usize & WORD_MASK;
+    let shift = offset * 8;
+
+    // Realign src
+    let mut src_aligned = (src as usize & !WORD_MASK) as *mut usize;
+    // XXX: Could this possibly be UB?
+    let mut prev_word = *src_aligned;
+
+    while dest_usize < dest_end {
+        src_aligned = src_aligned.add(1);
+        let cur_word = *src_aligned;
+        #[cfg(target_endian = "little")]
+        let resembled = prev_word >> shift | cur_word << (WORD_SIZE * 8 - shift);
+        #[cfg(target_endian = "big")]
+        let resembled = prev_word << shift | cur_word >> (WORD_SIZE * 8 - shift);
+        prev_word = cur_word;
+
+        *dest_usize = resembled;
+        dest_usize = dest_usize.add(1);
+    }
+}
+
+#[inline(always)]
+pub unsafe fn copy_forward(mut dest: *mut u8, mut src: *const u8, mut n: usize) {
+    if likely(n >= WORD_COPY_THRESHOLD) {
+        // Align dest
+        // Because of n >= 2 * WORD_SIZE, dst_misalignment < n
+        let dest_misalignment = (dest as usize).wrapping_neg() & WORD_MASK;
```

Comment:
From align_offset's doc:
> If it is not possible to align the pointer, the implementation returns usize::MAX. It is permissible for the implementation to always return usize::MAX. Only your algorithm's performance can depend on getting a usable offset here, not its correctness.

So that rules out the use of align_offset. Also note that this is the very hot path and I would prefer simple bit trcisk rather to rely on LLVM to optimise out a very complex function.

Reply:
Could you try benchmarking to see if there is overhead?

Reply:
It's not related to performance, it could simply not be used for correctness.
>  It is permissible for the implementation to always return usize::MAX.